### PR TITLE
Caching package touchups

### PR DESCRIPTION
**Pull request by kujenga**
- added some clarity and explicit use cases to the README
- Added `SetupContext` functionality for more reliable use of the `Context` cacher with child context objects.

### CODE REVIEW

File: caching/README.md [hunk 1]
```diff
@@ -47,17 +47,16 @@ func MyHandlerFunc(w http.ResponseWriter, r *http.Request) {
 
 	json.NewEncoder(w).Encode(myObj)
 }
-
 ```
 
 ## Instance
-The instance cacher is an ARC cache of fixed size. The intended use of this cache is for values that are relatively low-cost to retrieve and do not absolutely need to be synced across instances. Memcache has overhead in both time and cost as compared to this cacher (on the order of 1ms + storage costs), which may not be hugely significant, but is not necessary. Use of the instance cacher will also reduce memcache contention.
+The instance cacher is an [ARC cache](https://en.wikipedia.org/wiki/Adaptive_replacement_cache) of fixed size. Use this cache for values with relatively low retrieval cost (as an eviction may happen at any point) that do not need to be prefectly in sync across instances, and when you want cached data to be shared across multiple requests within a single instance. Use of the instance cacher in a `ByPrecedence` chain will also reduce memcache contention.
```
Comment (mammothbane) [inline caching/README.md]: *perfectly

File: caching/README.md [hunk 2]
```diff
@@ -47,17 +47,16 @@ func MyHandlerFunc(w http.ResponseWriter, r *http.Request) {
...
 
 ## Context
-The context cacher uses go maps stored in `Instance` to cache data for a given request. As such, it's possible that the entire cached content for a Context cacher could be evicted at one time, but this is unlikely given how often it is likely to be accessed. In any case, this cacher functions effectively identically to `Instance`, but has an `http.Handler` (`caching.CleanupContext`) that can be used as a middleware to purge cached values when a request completes.
+The context cacher builds on the `Instance` cache by using a Go map to store data for a given [`context.Context`](https://godoc.org/golang.org/x/net/context#Context). Because all data for a context is stored within a single map in the global `Instance` cache, all cached content for a given context could be evicted simultaneously. Use this cache in situations where you have data that is accesed many times within a request, but there is little benefit in having it cached across requests. This cacher functions effectively identically to `Instance`, but has an `http.Handler` (`caching.CleanupContext`) that can be used as a middleware to purge cached values when a request completes.
```
Comment (mammothbane) [inline caching/README.md]: *accessed

File: caching/context.go [hunk 3]
```diff
@@ -101,6 +111,19 @@ func (c *Context) getItem() *ctxMapItem {
 	return &elem
 }
 
+type contextKey int
+
+const cacheKey contextKey = iota
+
+// SetupContext returns a child context that contains a value which this package can use internally
+// to identity any other context derived from this one.
```
Comment (mammothbane) [inline caching/context.go]: *identify

**Comment by mammothbane** [comment]
otherwise, looks good

**Comment by kujenga** [comment]
Great! just added a little simplification of the memcache cacher lock as well.

**Comment by mammothbane** [comment]
nice, should be all set

**Pull request merged**